SmartDeviceLink/NSMutableDictionary+Store.m

```diff
@@ -24,32 +26,88 @@ - (nullable id)sdl_objectForName:(SDLRPCParameterName)name {
+    return [self sdl_objectsForName:name ofClass:NSString.class error:error];
+}
+
+#pragma mark - object
+
 - (nullable id)sdl_objectForName:(SDLRPCParameterName)name ofClass:(Class)classType {
-    NSObject *obj = [self sdl_objectForName:name];
+    NSError *error;
+    id object = [self sdl_objectForName:name ofClass:classType error:&error];
+    return error == nil ? object : nil;
+}
+
+- (nullable id)sdl_objectForName:(SDLRPCParameterName)name ofClass:(Class)classType error:(NSError * _Nullable __autoreleasing * _Nullable)error {
+    id obj = [self sdl_objectForName:name];
+
     if (obj == nil || [obj isKindOfClass:classType]) {
         return obj;
-    } else {
-        return [[classType alloc] initWithDictionary:(NSDictionary *)obj];
     }
+    // translate dictionaries to objects
+    if ([obj isKindOfClass:NSDictionary.class] && [classType instancesRespondToSelector:@selector(initWithDictionary:)]) {
+        obj = [[classType alloc] initWithDictionary:(NSDictionary *)obj];
+        // update store so that the object isn't created multiple times
+        [self sdl_setObject:obj forName:name];
+        return obj;
+    }
+    NSError *wrongObjectError = [NSError sdl_store_wrongObject:obj expectedType:classType];
+
+    NSAssert(NO, [[wrongObjectError userInfo] objectForKey:NSLocalizedFailureReasonErrorKey]);
+
+    if (error) {
+        *error = wrongObjectError;
+    }
+
+    return nil;
 }
 
-- (NSArray *)sdl_objectsForName:(SDLRPCParameterName)name ofClass:(Class)classType {
-    NSArray *array = [self sdl_objectForName:name];
-    if ([array isEqual:[NSNull null]]) {
-        return [NSMutableArray array];
-    } else if (array.count < 1 || [array.firstObject isMemberOfClass:classType]) {
+#pragma mark - objects
+
+- (nullable NSArray *)sdl_objectsForName:(SDLRPCParameterName)name ofClass:(Class)classType {
+    NSError *error;
+    id objects = [self sdl_objectsForName:name ofClass:classType error:&error];
+    return error == nil ? objects : nil;
+}
+
+- (nullable NSArray *)sdl_objectsForName:(SDLRPCParameterName)name ofClass:(Class)classType error:(NSError * _Nullable __autoreleasing * _Nullable)error {
+    NSArray *array = [self sdl_objectForName:name ofClass:NSArray.class error:error];
+    if (!array || array.count < 1 || [array.firstObject isKindOfClass:classType]) {
         // It's an array of the actual class type, just return
         return array;
-    } else {
+    }
+    if ([classType instancesRespondToSelector:@selector(initWithDictionary:)]) {
         // It's an array of dictionaries, make them into their class type
         NSMutableArray *newList = [NSMutableArray arrayWithCapacity:array.count];
         for (NSDictionary<NSString *, id> *dict in array) {
-            [newList addObject:[[classType alloc] initWithDictionary:dict]];
+            if ([dict isKindOfClass:NSDictionary.class]) {
```

Comment:
Crash won't be here because we don't send unrecognized selector to instance.
Your solution won't be work when array contain different object types `@[NSNumber, NSDictionary, NSObject]`.
So, I think we should keep it as is.